crates/driver/src/settlement_proposal.rs

```diff
@@ -0,0 +1,356 @@
+    /// Turns something like an indicative RFQ interaction into a simulatable `Interaction` by
+    /// finalizing the required call data.
+    async fn finalize(&self) -> Result<Arc<dyn Interaction>>;
+
+    fn metadata(&self) -> InteractionMetadata;
+}
+
+#[derive(Debug, Clone)]
+pub struct TradedOrder {
+    pub order: Order,
+    pub executed_amount: U256,
+}
+
+impl TradedOrder {
+    /// Computes buy_token_price of the order based on whether it's a user order or a liquidity order.
+    /// User orders are allowed to get surplus and therefore return the clearing price of the
+    /// buy_token whereas liquidity orders must not get surplus so they return their limit price.
+    fn buy_token_price(&self, clearing_prices: &HashMap<H160, U256>) -> Option<U256> {
+        match self.order.metadata.is_liquidity_order {
+            // liquidity orders have to be settled at their limit price
+            true => clearing_prices
+                .get(&self.order.data.sell_token)?
+                .checked_mul(self.order.data.sell_amount)?
+                .checked_div(self.order.data.buy_amount),
+            false => clearing_prices.get(&self.order.data.buy_token).cloned(),
+        }
+    }
+
+    /// Computes and returns the executed trade amounts given the settlements uniform clearing
+    /// prices. The computed fee will always be 0 because it does not matter for further
+    /// computations.
+    fn execution(&self, clearing_prices: &HashMap<H160, U256>) -> Option<TradeExecution> {
+        let sell_price = clearing_prices.get(&self.order.data.sell_token)?;
+        let buy_price = self.buy_token_price(clearing_prices)?;
+
+        let order = &self.order.data;
+        let (sell_amount, buy_amount) = match order.kind {
+            OrderKind::Sell => {
+                let sell_amount = self.executed_amount;
+                let buy_amount = sell_amount
+                    .checked_mul(*sell_price)?
+                    .checked_ceil_div(&buy_price)?;
+                (sell_amount, buy_amount)
+            }
+            OrderKind::Buy => {
+                let buy_amount = self.executed_amount;
+                let sell_amount = buy_amount
+                    .checked_mul(buy_price)?
+                    .checked_div(*sell_price)?;
+                (sell_amount, buy_amount)
+            }
+        };
+
+        Some(TradeExecution {
+            sell_token: order.sell_token,
+            buy_token: order.buy_token,
+            sell_amount,
+            buy_amount,
+            fee_amount: 0.into(),
+        })
+    }
+
+    /// Verifies that all our required properties are respected by this trade:
+    ///   - executed amounts make sense for order type (partially_fillable vs. fill-or-kill)
+    ///   - user order limit prices have been respected (TODO handle partially_fillable case)
+    fn sanity_check_trade_properties(&self, clearing_prices: &HashMap<H160, U256>) -> Result<()> {
```

Comment:
ubernit: Maybe we should only create a `TradeExecution` for valid `TradedOrders` and  `execution()` can return a `Result<TradeExecution>`.